File: java/leetcode/dfs/surrounded_regions.java
/**
 * 130. Surrounded Regions
 * Given a 2D board containing 'X' and 'O' (the letter O), capture all regions surrounded by 'X'.

A region is captured by flipping all 'O's into 'X's in that surrounded region.

Example:

X X X X
X O O X
X X O X
X O X X
After running your function, the board should be:

X X X X
X X X X
X X X X
X O X X
Explanation:

Surrounded regions shouldn’t be on the border, which means that any 'O' on the border of the board are not flipped to 'X'. Any 'O' that is not on the border and it is not connected to an 'O' on the border will be flipped to 'X'. Two cells are connected if they are adjacent cells connected horizontally or vertically.
 */


public class surrounded_regions {

    private int[][] direction = {{0,1}, {0,-1}, {1,0}, {-1,0}};
    private int m, n;

    public void solve(char[][] board) {
        if (board == null || board.length == 0) {
            return;
        }

        m = board.length;
        n = board[0].length;

        // first check O on boundary
        for (int i = 0; i < m; i++) {
            dfs(board, i, 0);
            dfs(board, i, n-1);
        }
        for (int i = 0; i < n; i++) {
            dfs(board, 0, i);
            dfs(board, m-1, i);
        }

        // next check and convert O not connect to boundary
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 'T') {
                    board[i][j] = 'O';
                } else if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
            }
        } 
    }

    private void dfs(char[][] board, int r, int c) {
        if (r < 0 || r >= m || c < 0 || c >= n || board[r][c] != 'O') {
            return;
        }
        board[r][c] = 'T';
        for (int[] d : direction) {
            dfs(board, r+d[0], c+d[1]);
        }
    }

}

File: golang/leetcode/dfs/surrounded_regions.go
package dfs

// 130. Surrounded Regions
/*
var (
	directions [][]int
	m          int
	n          int
)
*/

func solve(board [][]byte) {
	directions = [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
	if len(board) == 0 {
		return
	}
	m, n = len(board), len(board[0])

	for i := 0; i < m; i++ {
		regionTrack(board, i, 0)
		regionTrack(board, i, n-1)
	}
	for i := 0; i < n; i++ {
		regionTrack(board, 0, i)
		regionTrack(board, m-1, i)
	}

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if board[i][j] == 'T' {
				board[i][j] = 'O'
			} else if board[i][j] == 'O' {
				board[i][j] = 'X'
			}
		}
	}
}

func regionTrack(board [][]byte, r int, c int) {
	if r < 0 || r >= m || c < 0 || c >= n || board[r][c] != 'O' {
		return
	}
	board[r][c] = 'T'
	for _, d := range directions {
		regionTrack(board, r+d[0], c+d[1])
	}
}
